recipe/meta.yaml

```diff
@@ -6,54 +6,38 @@ package:
   version: {{ version }}
 
 source:
-  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  url: https://pypi.org/packages/source/{{ name[0] }}/{{ name }}/uproot-{{ version }}.tar.gz
   sha256: b8332c76baf3f327f84972a9b4d55a72c153f324331cdb85a1b0490bd14cf1b1
 
 build:
-  number: 0
+  noarch: python
```

Comment:
I originally had

```yaml
  skip: true # [py<38]
```

here but the linter failed with

> Non noarch packages should have python requirement without any version constraints.

and the CI was failing for macOS (for unknown reasons).

When I changed this back to

```yaml
  noarch: python
```

the linter is failing with 

> `noarch` packages can't have selectors. If the selectors are necessary, please remove `noarch: python`.

While conda-forge builds only for Python 3.9+ now, it seems like you should still want to ensure that the lower bound constraints are enforced to avoid people getting installations on the wrong CPython.

@conda-forge/help-python @chrisburr can you offer advice here?

Reply:
From https://conda-forge.org/docs/maintainer/knowledge_base/#noarch-python I'm confused about the defintion of "selectors" as it mentions:

> In order to qualify as a noarch python package, all of the following criteria must be fulfilled:
> 
> * No compiled extensions
> * No post-link or pre-link or pre-unlink scripts
> * No OS-specific build scripts
> * No python version specific requirements
> * No skips except for python version. If the recipe is py3 only, remove skip statement and add version constraint on python in `host` and `run` section.
> * `2to3` is not used
> * `scripts` argument in `setup.py` is not used
> * If `console_scripts` `entry_points` are defined in `setup.py` or `setup.cfg`, they are also [listed](https://conda.io/projects/conda-build/en/stable/resources/define-metadata.html#python-entry-points) in the `build` section of `meta.yaml`
> * No activate scripts

but then also mentions

> [!NOTE] 
> While `noarch: python` does not work with selectors, it does work with version constraints. `skip: True # [py2k]` can be replaced with a constrained python version in the `host` and `run` subsections: say `python >=3` instead of just `python`.

This seems to be exactly what we're doing with

```yaml
requirements:
  host:
    - python >=3.8
    - hatchling >1.10.0
    - hatch-vcs
    - pip
  run:
    - python >=3.8
```

so I'm confused why the linter is rejecting this.

Reply:
Oh. From the conda-forge docs FAQ [Why does conda-build ignore the py37 selector in meta.yaml?](https://conda-forge.org/docs/maintainer/maintainer_faq/#why-does-conda-build-ignore-the-py37-selector-in-metayaml) (which links to Selectors in conda-build documentation ([Preprocessing selectors](https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#preprocessing-selectors))) I now realize that "selectors" are the

```
# [<selector>]
```

format.

So the selector the linter is complaining about is the `# [py<311]` in

```yaml
  run:
    - python >=3.8
    - awkward >=2.4.6
    - cramjam >=2.5.0
    - python-xxhash
    - numpy
    - fsspec
    - packaging
    - typing_extensions >=4.1.0  # [py<311]
```

So I guess that needs to be removed even and `typing_extensions` will get installed for everything?